src/Retry.php

```diff
@@ -21,80 +21,57 @@ class Retry
      *
      * @var callable
      */
-    private $callable;
+    private $operation;
 
     /**
      * Maximum number of retries.
      *
-     * @var integer
+     * @var int
      */
     private $maxRetries;
 
     /**
-     * Delay between retries in milliseconds.
+     * Actual number of retries.
      *
-     * @var integer
+     * @var int|null
      */
-    private $retryDelay;
+    private $retries;
 
     /**
-     * Actual number of retries.
+     * Exceptions to catch and retry on.
      *
-     * @var integer|null
+     * @var string[]
      */
-    private $retries;
+    private $exceptions = [];
 
     /**
-     * Exceptions to catch and retry on.
+     * Callback when an exception is caught.
      *
-     * @var array
+     * @var callable
      */
-    private $exceptions = array();
+    private $exceptionCallback;
 
     /**
-     * Constructor to wrap a callable.
+     * Constructor to wrap a callable operation.
      *
-     * @param callable      $callable   The operation to execute that can be retried on failure.
-     * @param string|array  $exceptions Exceptions to catch and retry
-     * @param integer       $maxRetries Maximum number of retries.
-     * @param callable|null $retryDelay Delay between retries in milliseconds
+     * @param callable        $operation         The operation to execute that should be retried on failure
+     * @param string|string[] $exceptions        Exceptions to catch and retry on (by default every exception)
+     * @param int             $maxRetries        Maximum number of retries
+     * @param callable|null   $exceptionCallback The callback to execute when an exception is caught and the operation is about to be retried.
+     *                                           By default, it delays retries by 300 milliseconds. The callback receives the exception as parameter.
      */
-    public function __construct(callable $callable, $exceptions = 'Exception', $maxRetries = 3, callable $retryDelay = null)
+    public function __construct(callable $operation, $exceptions = 'Exception', $maxRetries = 2, callable $exceptionCallback = null)
```

Comment:
in theory $exceptions and $maxRetries could also be represented as an exception callback.
so we could just have an array of callbacks.